File: src/slack_stealth_mcp/tools/get_context.py
"""Tool to get conversation context around a message."""

from __future__ import annotations

from typing import Any

from ..workspace import WorkspaceManager


async def get_context(
    manager: WorkspaceManager,
    channel: str,
    message_ts: str | None = None,
    context_size: int = 10,
    workspace: str | None = None,
) -> dict[str, Any]:
    """Get messages from a conversation, optionally around a specific message.

    This does NOT mark messages as read.

    Args:
        manager: Workspace manager
        channel: Channel ID (C..., D..., G...)
        message_ts: Optional message timestamp to get context around
        context_size: Number of messages to fetch (default 10)
        workspace: Workspace name (uses default if not specified)

    Returns:
        Dictionary with messages and conversation info
    """
    client = manager.get_client(workspace)

    # Check if this is a thread request (message_ts points to a thread parent)
    is_thread = False
    messages = []

    if message_ts:
        # Try to get thread replies first
        try:
            thread_messages = await client.get_thread_replies(
                channel=channel,
                thread_ts=message_ts,
                limit=context_size,
            )
            if len(thread_messages) > 1:
                # It's a thread with replies
                is_thread = True
                messages = thread_messages
        except Exception:
            pass

    if not messages:
        # Get channel history
        if message_ts:
            # Get messages around the specified timestamp
            # First get messages before (including target)
            messages = await client.get_conversation_history(
                channel=channel,
                latest=message_ts,
                limit=context_size // 2 + 1,
                inclusive=True,
            )
            # Then get messages after
            after_messages = await client.get_conversation_history(
                channel=channel,
                oldest=message_ts,
                limit=context_size // 2,
                inclusive=False,
            )
            # Combine (after are newest first, messages are also newest first)
            messages = after_messages + messages
        else:
            # Just get recent messages
            messages = await client.get_conversation_history(
                channel=channel,
                limit=context_size,
            )

    # Resolve user names
    formatted_messages = []
    for msg in reversed(messages):  # Reverse to show oldest first
        user_name = "Unknown"
        if msg.user:
            user_name = await client.resolve_user_name(msg.user)

        formatted = {
            "ts": msg.ts,
            "user": user_name,
            "text": msg.text,
            "time": msg.timestamp.strftime("%Y-%m-%d %H:%M:%S"),
        }

        if msg.thread_ts and msg.thread_ts != msg.ts:
            formatted["thread_ts"] = msg.thread_ts
            formatted["is_reply"] = True

        if msg.reply_count:
            formatted["reply_count"] = msg.reply_count

        if msg.reactions:
            formatted["reactions"] = [
                f":{r['name']}:{r.get('count', 1)}" for r in msg.reactions
            ]

        formatted_messages.append(formatted)

    # Get conversation info for context
    try:
        conv = await client.get_conversation_info(channel)
        channel_name = conv.display_name
        channel_type = conv.conversation_type
    except Exception:
        channel_name = channel
        channel_type = "unknown"

    result: dict[str, Any] = {
        "channel": channel_name,
        "channel_id": channel,
        "channel_type": channel_type,
        "messages": formatted_messages,
        "message_count": len(formatted_messages),
    }

    if is_thread:
        result["is_thread"] = True
        result["thread_ts"] = message_ts

    return result

File: packages/cloudflare/src/tools/get-context.ts
/**
 * Get Context Tool - Fetch conversation messages for context
 */

import { SlackClient } from "../slack-client";
import { tsToDate, getConversationType } from "../types/slack";

interface GetContextOptions {
  channel: string;
  messageTs?: string;
  contextSize?: number;
}

interface FormattedMessage {
  ts: string;
  user: string;
  text: string;
  time: string;
  thread_ts?: string;
  is_reply?: boolean;
  reply_count?: number;
  reactions?: string[];
}

interface GetContextResult {
  channel: string;
  channel_id: string;
  channel_type: string;
  messages: FormattedMessage[];
  message_count: number;
  is_thread?: boolean;
  thread_ts?: string;
}

/**
 * Get messages from a conversation, optionally around a specific message.
 * This does NOT mark messages as read.
 */
export async function getContext(
  client: SlackClient,
  options: GetContextOptions
): Promise<GetContextResult> {
  const { channel, messageTs, contextSize = 10 } = options;

  // Check if this is a thread request
  let isThread = false;
  let messages: Awaited<ReturnType<typeof client.getConversationHistory>> = [];

  if (messageTs) {
    // Try to get thread replies first
    try {
      const threadMessages = await client.getThreadReplies(
        channel,
        messageTs,
        contextSize
      );
      if (threadMessages.length > 1) {
        // It's a thread with replies
        isThread = true;
        messages = threadMessages;
      }
    } catch {
      // Not a thread or error - will fetch channel history
    }
  }

  if (messages.length === 0) {
    // Get channel history
    if (messageTs) {
      // Get messages around the specified timestamp
      // First get messages before (including target)
      const beforeMessages = await client.getConversationHistory(channel, {
        latest: messageTs,
        limit: Math.floor(contextSize / 2) + 1,
        inclusive: true,
      });

      // Then get messages after
      const afterMessages = await client.getConversationHistory(channel, {
        oldest: messageTs,
        limit: Math.floor(contextSize / 2),
        inclusive: false,
      });

      // Combine (after are newest first, messages are also newest first)
      messages = [...afterMessages, ...beforeMessages];
    } else {
      // Just get recent messages
      messages = await client.getConversationHistory(channel, {
        limit: contextSize,
      });
    }
  }

  // Format messages - use cached user names
  const formattedMessages: FormattedMessage[] = [];
  for (const msg of messages.slice().reverse()) {
    // Reverse to show oldest first
    let userName = "Unknown";
    if (msg.user) {
      const cached = client.getCachedUser(msg.user);
      userName = cached
        ? cached.display_name || cached.real_name || cached.name
        : msg.user.substring(0, 8) + "...";
    }

    const formatted: FormattedMessage = {
      ts: msg.ts,
      user: userName,
      text: msg.text,
      time: tsToDate(msg.ts).toISOString().replace("T", " ").substring(0, 19),
    };

    if (msg.thread_ts && msg.thread_ts !== msg.ts) {
      formatted.thread_ts = msg.thread_ts;
      formatted.is_reply = true;
    }

    if (msg.reply_count) {
      formatted.reply_count = msg.reply_count;
    }

    if (msg.reactions && msg.reactions.length > 0) {
      formatted.reactions = msg.reactions.map(
        (r) => `:${r.name}:${r.count > 1 ? r.count : ""}`
      );
    }

    formattedMessages.push(formatted);
  }

  // Get conversation info for context
  let channelName = channel;
  let channelType = "unknown";
  try {
    const conv = await client.getConversationInfo(channel);
    channelName = conv.name || channel;
    channelType = getConversationType(conv);
  } catch {
    // Ignore - use defaults
  }

  const result: GetContextResult = {
    channel: channelName,
    channel_id: channel,
    channel_type: channelType,
    messages: formattedMessages,
    message_count: formattedMessages.length,
  };

  if (isThread) {
    result.is_thread = true;
    result.thread_ts = messageTs;
  }

  return result;
}
